src/main/java/org/icatproject/core/manager/search/SearchManager.java

```diff
@@ -0,0 +1,817 @@
+						String line;
+						Set<String> ids = new HashSet<>();
+						while ((line = reader.readLine()) != null) {
+							if (ids.add(line)) { // True if id not yet encountered
+								String query = "SELECT e FROM " + entityName + " e WHERE e.id = " + line;
+								try {
+									EntityBaseBean entity = entityManager.createQuery(query, klass).getSingleResult();
+									updateDocument(entity);
+								} catch (Exception e) {
+									logger.error("{} with id {} not found, continue", entityName, line);
+								}
+							}
+						}
+						file.delete();
+						logger.info(entityName + " aggregations performed");
+					} catch (IOException e) {
+						logger.error("Problems reading from {} : {}", file, e.getMessage());
+					} catch (Throwable e) {
+						logger.error("Something unexpected happened " + e.getClass() + " " + e.getMessage());
+					}
+					logger.debug("finish processing");
+				}
+			}
+		}
+	}
+
+	private enum PopState {
+		STOPPING, STOPPED
+	}
+
+	/**
+	 * Holds relevant values for a Populate thread.
+	 */
+	private class PopulateBucket {
+		private Long minId;
+		private Long maxId;
+		private boolean delete;
+
+		public PopulateBucket(Long minId, Long maxId, boolean delete) {
+			this.minId = minId;
+			this.maxId = maxId;
+			this.delete = delete;
+		}
+	}
+
+	public class PopulateThread extends Thread {
+
+		private EntityManager manager;
+		private EntityManagerFactory entityManagerFactory;
+
+		public PopulateThread(EntityManagerFactory entityManagerFactory) {
+			this.entityManagerFactory = entityManagerFactory;
+			manager = entityManagerFactory.createEntityManager();
+			logger.info("Start new populate thread");
+		}
+
+		@Override
+		public void run() {
+
+			try {
+				while (!populateMap.isEmpty()) {
+
+					populatingClassEntry = populateMap.firstEntry();
+
+					if (populatingClassEntry != null) {
+						PopulateBucket bucket = populatingClassEntry.getValue();
+						Long start = bucket.minId != null && bucket.minId > 0 ? bucket.minId : 0;
+						searchApi.lock(populatingClassEntry.getKey(), bucket.minId, bucket.maxId, bucket.delete);
+
+						logger.info("Search engine populating " + populatingClassEntry);
+
+						CompletionService<Long> threads = new ExecutorCompletionService<>(populateExecutor);
+						SortedSet<Long> tasks = new ConcurrentSkipListSet<>();
+
+						while (true) {
+
+							if (popState == PopState.STOPPING) {
+								logger.info("PopulateThread stopping as flag was set");
+								break;
+							}
+							/* Get next block of ids */
+							String query = "SELECT e.id from " + populatingClassEntry.getKey() + " e";
+							if (bucket.maxId != null) {
+								// Add 1 from lower limit to get a half interval
+								query += " WHERE e.id BETWEEN " + (start + 1) + " AND " + (bucket.maxId);
+							} else {
+								query += " WHERE e.id > " + start;
+							}
+							query += " ORDER BY e.id";
+							List<Long> ids = manager
+									.createQuery(query, Long.class)
```

Comment:
The inputs are supplied by the user (this was the case before this change, I've just added the option to go "between" two values in addition to greater than min). The following checks are (still) in place:

1. Only a root user can call this:
https://github.com/icatproject/icat.server/blob/de8fd2d73b54de68c52c5e8479750edcb4669392/src/main/java/org/icatproject/exposed/ICATRest.java#L1626-L1631
2. Entity name has to be the name of an entity, or we throw:
https://github.com/icatproject/icat.server/blob/de8fd2d73b54de68c52c5e8479750edcb4669392/src/main/java/org/icatproject/core/manager/EntityBeanManager.java#L1708-L1718
3. `minId` and `maxId` have to be longs, so (I'm assuming) can't be used to do anything malicious

In your opinion, is this sufficient? I thought it would be given we were doing it before, but if not let me know what improvement's you'd recommend here.

Reply:
Thanks for this. Yes, I think that is sufficient. Was just double checking.